File: src/task_service.cpp
#include "task_service.h"
#include "date_utils.h"
#include <iostream>
#include <algorithm>

TaskService::TaskService(TaskStore& store) : store_(store) {}

// =============================================================================
// Create — primary method for the "Create a New Task" story.
// All creation-related business rules are centralized here.
// =============================================================================
std::optional<Task> TaskService::createTask(
    const std::string& title,
    Priority priority,
    const std::string& description,
    const std::string& due_date,
    const std::string& assignee)
{
    // Validate title: must be non-empty after trimming whitespace
    std::string trimmed = title;
    trimmed.erase(0, trimmed.find_first_not_of(" \t\n\r"));
    trimmed.erase(trimmed.find_last_not_of(" \t\n\r") + 1);
    if (trimmed.empty()) {
        std::cerr << "Title cannot be empty." << std::endl;
        return std::nullopt;
    }

    // Validate due date if provided
    if (!due_date.empty() && !is_valid_date(due_date)) {
        std::cerr << "Invalid date format — due date not set." << std::endl;
        return std::nullopt;
    }

    // Build the task
    int id = store_.next_id();
    Task task(id, title, today_as_string());
    task.set_priority(priority);
    task.set_description(description);
    task.set_due_date(due_date);
    task.set_assignee(assignee);

    return store_.save(task);
}

// =============================================================================
// Read
// =============================================================================
std::optional<Task> TaskService::getTask(int id) const {
    return store_.find_by_id(id);
}

std::vector<Task> TaskService::getAllTasks() const {
    return store_.find_all();
}

std::vector<Task> TaskService::getPendingTasks() const {
    return store_.find_by_status(TaskStatus::TODO);
}

std::vector<Task> TaskService::getTasksSortedByPriority() const {
    auto tasks = store_.find_all();
    std::sort(tasks.begin(), tasks.end(), [](const Task& a, const Task& b) {
        return static_cast<int>(a.get_priority()) > static_cast<int>(b.get_priority());
    });
    return tasks;
}

std::vector<Task> TaskService::getOverdueTasks() const {
    auto all = store_.find_all();
    std::vector<Task> overdue;
    std::copy_if(all.begin(), all.end(), std::back_inserter(overdue),
                 [](const Task& t) { return t.is_overdue(); });
    return overdue;
}

TaskSummary TaskService::getSummary() const {
    auto all = store_.find_all();
    TaskSummary summary;
    summary.total = static_cast<int>(all.size());
    for (const auto& t : all) {
        switch (t.get_status()) {
            case TaskStatus::TODO:        ++summary.todo; break;
            case TaskStatus::IN_PROGRESS: ++summary.in_progress; break;
            case TaskStatus::DONE:        ++summary.done; break;
            case TaskStatus::CANCELLED:   ++summary.cancelled; break;
        }
        if (t.is_overdue()) {
            ++summary.overdue;
        }
    }
    return summary;
}

// =============================================================================
// Update — state transitions via the store's update_status helper
// =============================================================================
TransitionResult TaskService::startTask(int id) {
    return store_.update_status(id, [](Task& t) { return t.start_progress(); });
}

TransitionResult TaskService::completeTask(int id) {
    return store_.update_status(id, [](Task& t) { return t.complete(); });
}

TransitionResult TaskService::cancelTask(int id) {
    return store_.update_status(id, [](Task& t) { return t.cancel(); });
}

bool TaskService::assignTask(int id, const std::string& assignee) {
    return store_.update(id, [&assignee](Task& t) { t.set_assignee(assignee); });
}

bool TaskService::updatePriority(int id, Priority priority) {
    return store_.update(id, [priority](Task& t) { t.set_priority(priority); });
}

// =============================================================================
// Delete
// =============================================================================
bool TaskService::deleteTask(int id) {
    return store_.delete_task(id);
}

File: src/main/java/com/demo/TaskService.java
package com.demo;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

public class TaskService {
    private final TaskRepository repository;

    public TaskService(TaskRepository repository) {
        this.repository = repository;
    }

    public Task createTask(String title) {
        return createTask(title, Priority.MEDIUM, null, null);
    }

    public Task createTask(String title, Priority priority, String description, LocalDate dueDate) {
        Task task = new Task(repository.nextId(), title);
        task.setPriority(priority);
        task.setDescription(description);
        task.setDueDate(dueDate);
        return repository.save(task);
    }

    public Optional<Task> getTask(int id) {
        return repository.findById(id);
    }

    public List<Task> getAllTasks() {
        return repository.findAll();
    }

    public List<Task> getPendingTasks() {
        return repository.findByStatus(TaskStatus.TODO);
    }

    public List<Task> getInProgressTasks() {
        return repository.findByStatus(TaskStatus.IN_PROGRESS);
    }

    public List<Task> getCompletedTasks() {
        return repository.findByStatus(TaskStatus.DONE);
    }

    public List<Task> getOverdueTasks() {
        return repository.findAll().stream()
                .filter(Task::isOverdue)
                .collect(Collectors.toList());
    }

    public List<Task> getCancelledTasks() {
        return repository.findByStatus(TaskStatus.CANCELLED);
    }

    public List<Task> getFilteredTasks(String filter) {
        switch (filter) {
            case "all":
                return getTasksSortedByPriority();
            case "todo":
                return repository.findByStatus(TaskStatus.TODO);
            case "in_progress":
                return repository.findByStatus(TaskStatus.IN_PROGRESS);
            case "done":
                return repository.findByStatus(TaskStatus.DONE);
            case "cancelled":
                return repository.findByStatus(TaskStatus.CANCELLED);
            case "overdue":
                return repository.findAll().stream()
                        .filter(Task::isOverdue)
                        .collect(Collectors.toList());
            default:
                return new ArrayList<>();
        }
    }

    public List<Task> getTasksByPriority(Priority priority) {
        return repository.findByPriority(priority);
    }

    public List<Task> getTasksSortedByPriority() {
        return repository.findAll().stream()
                .sorted(Comparator.comparing(Task::getPriority).reversed())
                .collect(Collectors.toList());
    }

    public List<Task> getTasksSortedByDueDate() {
        Comparator<Task> dueDateComparator = (left, right) -> {
            boolean leftHas = left.getDueDate() != null;
            boolean rightHas = right.getDueDate() != null;
            if (leftHas && rightHas) {
                return left.getDueDate().compareTo(right.getDueDate());
            }
            if (leftHas) {
                return -1;
            }
            if (rightHas) {
                return 1;
            }
            return Integer.compare(left.getId(), right.getId());
        };
        return repository.findAll().stream()
                .sorted(dueDateComparator)
                .collect(Collectors.toList());
    }

    public List<Task> getTasksSortedByCreationOrder() {
        return repository.findAll().stream()
                .sorted(Comparator.comparingInt(Task::getId))
                .collect(Collectors.toList());
    }

    public boolean startTask(int id) {
        return repository.findById(id).map(t -> {
            t.startProgress();
            return true;
        }).orElse(false);
    }

    public boolean completeTask(int id) {
        return repository.findById(id).map(t -> {
            t.complete();
            return true;
        }).orElse(false);
    }

    public boolean cancelTask(int id) {
        return repository.findById(id).map(t -> {
            t.cancel();
            return true;
        }).orElse(false);
    }

    public boolean assignTask(int id, String assignee) {
        return repository.findById(id).map(t -> {
            t.setAssignee(assignee);
            return true;
        }).orElse(false);
    }

    public boolean updatePriority(int id, Priority priority) {
        return repository.findById(id).map(t -> {
            t.setPriority(priority);
            return true;
        }).orElse(false);
    }

    public boolean deleteTask(int id) {
        return repository.delete(id);
    }

    public TaskSummary getSummary() {
        List<Task> all = repository.findAll();
        long todo = all.stream().filter(t -> t.getStatus() == TaskStatus.TODO).count();
        long inProgress = all.stream().filter(t -> t.getStatus() == TaskStatus.IN_PROGRESS).count();
        long done = all.stream().filter(t -> t.getStatus() == TaskStatus.DONE).count();
        long cancelled = all.stream().filter(t -> t.getStatus() == TaskStatus.CANCELLED).count();
        long overdue = all.stream().filter(Task::isOverdue).count();
        return new TaskSummary(all.size(), todo, inProgress, done, cancelled, overdue);
    }
}
